baseline/moc.py

```diff
@@ -0,0 +1,127 @@
+import logging
+from typing import Dict, Mapping, Tuple
+
+import mlflow
+import pandas as pd
+import typer
+from sklearn.metrics import mean_squared_error
+
+from ica.ica import run as run_ica
+from lib.config import AppConfig
+from lib.data_handling import CompositionData
+from lib.reproduction import major_oxides, weighted_sum_oxide_percentages
+from PLS_SM.full_flow import full_run as run_pls
+
+logger = logging.getLogger(__name__)
+
+app = typer.Typer()
+
+
+def run_ica_pls_for_predictions_on_full_data() -> Tuple[pd.DataFrame, pd.DataFrame]:
+    logger.info("Getting predictions for PLS-SM")
+    pls_tar_pred: pd.DataFrame = run_pls()
+
+    logger.info("Getting predictions for ICA")
+    ica_tar_pred: pd.DataFrame = run_ica()
+
+    logger.info("Finished getting predictions for PLS-SM and ICA")
+
+    return pls_tar_pred, ica_tar_pred
+
+
+def align_predictions(pls_tar_pred, ica_tar_pred) -> pd.DataFrame:
+    logger.info("Aligning predictions")
+    ica_tar_pred["Sample Name"] = ica_tar_pred["Sample Name"].apply(
+        lambda x: x.split("_")[0]
+    )
+
+    ica_tar_pred.set_index("Sample Name", inplace=True)
+    pls_tar_pred.set_index("Sample Name", inplace=True)
+
+    return pd.merge(
+        ica_tar_pred, pls_tar_pred, left_index=True, right_index=True, how="inner"
+    )
+
+
+def make_moc_predictions(
+    pls_predictions: pd.DataFrame,
+    ica_predictions: pd.DataFrame,
+    blend_ratios: Mapping[str, Mapping[str, float | int]],
+) -> pd.DataFrame:
+    logger.info("Making MOC predictions")
+    merged_df = align_predictions(
+        ica_tar_pred=ica_predictions, pls_tar_pred=pls_predictions
+    )
+
+    moc_predictions = pd.DataFrame()
+
+    for oxide, ratio in blend_ratios.items():
+        w_ica = ratio["ICA"] / 100
+        w_pls_sm = ratio["PLS1-SM"] / 100
+        moc_predictions[oxide] = (
+            merged_df[oxide + "_ICA"] * w_ica + merged_df[oxide + "_PLS_SM"] * w_pls_sm
+        )
+
+    moc_predictions["Sample Name"] = merged_df["Sample Name"]
+    moc_predictions["ID"] = merged_df.index
+
+    return moc_predictions
+
+
+def merge_predictions_with_actual(moc_predictions: pd.DataFrame) -> pd.DataFrame:
+    logger.info("Merging predictions with actual data")
+    cd = CompositionData(composition_data_loc=AppConfig().composition_data_path)
+
+    merged_data = pd.DataFrame()
+
+    for index, row in moc_predictions.iterrows():
+        actual_data = cd.get_composition_for_sample(row["Sample Name"])
+
+        if not actual_data.empty:
```

Comment:
Should this not just have an early return with an error if actual data is empty and then do the following loop if its not empty?

Reply:
Can `continue` but it's not necessary to do an early return

Reply:
But it's the same as what's here now practically